File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/forum/AnnounceFragment.java
package forpdateam.ru.forpda.ui.fragments.forum;

import android.annotation.SuppressLint;
import android.app.SearchManager;
import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatImageButton;
import android.support.v7.widget.SearchView;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;

import java.util.ArrayList;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.R;
import forpdateam.ru.forpda.api.forum.models.Announce;
import forpdateam.ru.forpda.apirx.RxApi;
import forpdateam.ru.forpda.common.rx.Subscriber;
import forpdateam.ru.forpda.common.webview.CustomWebChromeClient;
import forpdateam.ru.forpda.common.webview.CustomWebViewClient;
import forpdateam.ru.forpda.ui.activities.MainActivity;
import forpdateam.ru.forpda.ui.fragments.TabFragment;
import forpdateam.ru.forpda.ui.views.ExtendedWebView;

/**
 * Created by radiationx on 16.10.17.
 */

public class AnnounceFragment extends TabFragment {
    public final static String ARG_ANNOUNCE_ID = "ARG_ANNOUNCE_ID";
    public final static String ARG_FORUM_ID = "ARG_FORUM_ID";
    public final static String JS_INTERFACE = "IAnnounce";
    private ExtendedWebView webView;
    private Subscriber<Announce> mainSubscriber = new Subscriber<>(this);
    private int searchViewTag = 0;
    private int id = 0;
    private int forumId = 0;


    public AnnounceFragment() {
        configuration.setDefaultTitle("Объявление");
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            id = getArguments().getInt(ARG_ANNOUNCE_ID);
            forumId = getArguments().getInt(ARG_FORUM_ID);
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        webView = getMainActivity().getWebViewsProvider().pull(getContext());
        attachWebView(webView);
        fragmentContent.addView(webView);
        return view;
    }

    @SuppressLint("JavascriptInterface")
    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewsReady();
        webView.addJavascriptInterface(this, JS_INTERFACE);
        webView.setWebViewClient(new CustomWebViewClient());
        webView.setWebChromeClient(new CustomWebChromeClient());
        webView.setJsLifeCycleListener(new ExtendedWebView.JsLifeCycleListener() {
            @Override
            public void onDomContentComplete(ArrayList<String> actions) {
                setRefreshing(false);
            }

            @Override
            public void onPageComplete(ArrayList<String> actions) {

            }
        });
    }

    @Override
    protected void addBaseToolbarMenu() {
        super.addBaseToolbarMenu();
        addSearchOnPageItem(getMenu());
    }

    @Override
    public boolean loadData() {
        if (!super.loadData()) {
            return false;
        }
        setRefreshing(true);
        mainSubscriber.subscribe(RxApi.Forum().getAnnounce(id, forumId, true), this::onLoad, new Announce(), view1 -> loadData());
        return true;
    }

    private void onLoad(Announce announce) {
        setTitle(announce.getTitle());
        webView.loadDataWithBaseURL("https://4pda.ru/forum/", announce.getHtml(), "text/html", "utf-8", null);
        new Handler().postDelayed(() -> {
            if (isAdded())
                setRefreshing(false);
        }, 1000);
    }

    private void addSearchOnPageItem(Menu menu) {
        toolbar.inflateMenu(R.menu.theme_search_menu);
        MenuItem searchOnPageMenuItem = menu.findItem(R.id.action_search);
        searchOnPageMenuItem.setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS);
        SearchView searchView = (SearchView) searchOnPageMenuItem.getActionView();
        searchView.setTag(searchViewTag);

        searchView.setOnSearchClickListener(v -> {
            if (searchView.getTag().equals(searchViewTag)) {
                ImageView searchClose = (ImageView) searchView.findViewById(android.support.v7.appcompat.R.id.search_close_btn);
                if (searchClose != null)
                    ((ViewGroup) searchClose.getParent()).removeView(searchClose);

                ViewGroup.LayoutParams navButtonsParams = new ViewGroup.LayoutParams(App.px48, App.px48);
                TypedValue outValue = new TypedValue();
                getContext().getTheme().resolveAttribute(android.R.attr.actionBarItemBackground, outValue, true);

                AppCompatImageButton btnNext = new AppCompatImageButton(searchView.getContext());
                btnNext.setImageDrawable(App.getVecDrawable(getContext(), R.drawable.ic_toolbar_search_next));
                btnNext.setBackgroundResource(outValue.resourceId);

                AppCompatImageButton btnPrev = new AppCompatImageButton(searchView.getContext());
                btnPrev.setImageDrawable(App.getVecDrawable(getContext(), R.drawable.ic_toolbar_search_prev));
                btnPrev.setBackgroundResource(outValue.resourceId);

                ((LinearLayout) searchView.getChildAt(0)).addView(btnPrev, navButtonsParams);
                ((LinearLayout) searchView.getChildAt(0)).addView(btnNext, navButtonsParams);

                btnNext.setOnClickListener(v1 -> findNext(true));
                btnPrev.setOnClickListener(v1 -> findNext(false));
                searchViewTag++;
            }
        });

        SearchManager searchManager = (SearchManager) getMainActivity().getSystemService(Context.SEARCH_SERVICE);
        if (null != searchManager) {
            searchView.setSearchableInfo(searchManager.getSearchableInfo(getMainActivity().getComponentName()));
        }

        searchView.setIconifiedByDefault(true);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                findText(newText);
                return false;
            }
        });
    }

    protected void findNext(boolean next) {
        webView.findNext(next);
    }

    protected void findText(String text) {
        webView.findAllAsync(text);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (webView != null) {
            webView.endWork();
            ((MainActivity) getActivity()).getWebViewsProvider().push(webView);
        }
    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/other/AnnounceFragment.kt
package forpdateam.ru.forpda.ui.fragments.other

import android.annotation.SuppressLint
import android.app.SearchManager
import android.content.Context
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.appcompat.widget.AppCompatImageButton
import androidx.appcompat.widget.SearchView
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.common.Utils
import forpdateam.ru.forpda.common.webview.CustomWebChromeClient
import forpdateam.ru.forpda.common.webview.CustomWebViewClient
import forpdateam.ru.forpda.common.webview.DialogsHelper
import forpdateam.ru.forpda.entity.remote.forum.Announce
import forpdateam.ru.forpda.extensions.getDimenPx
import forpdateam.ru.forpda.extensions.getDrawableResAttr
import forpdateam.ru.forpda.extensions.quillMoxyPresenter
import forpdateam.ru.forpda.presentation.ILinkHandler
import forpdateam.ru.forpda.presentation.ISystemLinkHandler
import forpdateam.ru.forpda.presentation.TabRouter
import forpdateam.ru.forpda.presentation.announce.AnnouncePresenter
import forpdateam.ru.forpda.presentation.announce.AnnounceView
import forpdateam.ru.forpda.ui.fragments.TabFragment
import forpdateam.ru.forpda.ui.fragments.TabTopScroller
import forpdateam.ru.forpda.ui.fragments.WebViewTopScroller
import forpdateam.ru.forpda.ui.views.ExtendedWebView
import ru.radiationx.quill.inject

/**
 * Created by radiationx on 16.10.17.
 */

class AnnounceFragment : TabFragment(), AnnounceView, TabTopScroller {

    private var searchViewTag = 0
    private lateinit var webView: ExtendedWebView
    private lateinit var topScroller: WebViewTopScroller

    private val utils by inject<Utils>()
    private val linkHandler by inject<ILinkHandler>()
    private val systemLinkHandler by inject<ISystemLinkHandler>()
    private val router by inject<TabRouter>()
    private val webViewClient by inject<CustomWebViewClient>()
    private val presenter by quillMoxyPresenter<AnnouncePresenter>()

    init {
        configuration.defaultTitle = "Объявление"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.apply {
            presenter.id = getInt(ARG_ANNOUNCE_ID)
            presenter.forumId = getInt(ARG_FORUM_ID)
        }
    }

    @SuppressLint("JavascriptInterface")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        webView = ExtendedWebView(requireContext())
        webView.setDialogsHelper(
            DialogsHelper(
                context = webView.context,
                linkHandler = linkHandler,
                systemLinkHandler = systemLinkHandler,
                utils = utils,
                router = router
            )
        )
        attachWebView(webView)
        fragmentContent.addView(webView)

        webView.addJavascriptInterface(this, JS_INTERFACE)
        webView.webViewClient = webViewClient
        webView.webChromeClient = CustomWebChromeClient()
        webView.setJsLifeCycleListener(object : ExtendedWebView.JsLifeCycleListener {
            override fun onDomContentComplete(actions: ArrayList<String>) {
                setRefreshing(false)
            }

            override fun onPageComplete(actions: ArrayList<String>) {

            }
        })
        topScroller = WebViewTopScroller(webView, appBarLayout)
    }

    override fun toggleScrollTop() {
        topScroller.toggleScrollTop()
    }

    override fun addBaseToolbarMenu(menu: Menu) {
        super.addBaseToolbarMenu(menu)
        addSearchOnPageItem(menu)
    }

    override fun showData(data: Announce) {
        setTitle(data.title)
        webView.loadDataWithBaseURL("https://4pda.to/forum/", data.html, "text/html", "utf-8", null)
    }

    override fun setStyleType(type: String) {
        webView.evalJs("changeStyleType(\"$type\")")
    }

    private fun addSearchOnPageItem(menu: Menu) {
        toolbar.inflateMenu(R.menu.theme_search_menu)
        val searchOnPageMenuItem = menu.findItem(R.id.action_search)
        searchOnPageMenuItem.setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS)
        val searchView = searchOnPageMenuItem.actionView as SearchView
        searchView.tag = searchViewTag

        searchView.setOnSearchClickListener { _ ->
            if (searchView.tag == searchViewTag) {
                val searchClose =
                    searchView.findViewById<View>(androidx.appcompat.R.id.search_close_btn) as ImageView?
                if (searchClose != null)
                    (searchClose.parent as ViewGroup).removeView(searchClose)

                val navButtonsParams = ViewGroup.LayoutParams(
                    searchView.context.getDimenPx(R.dimen.dp48),
                    searchView.context.getDimenPx(R.dimen.dp48)
                )

                val backgroundRes = requireContext().getDrawableResAttr(android.R.attr.actionBarItemBackground)

                val btnNext = AppCompatImageButton(searchView.context)
                btnNext.setImageResource(R.drawable.ic_toolbar_search_next)
                btnNext.setBackgroundResource(backgroundRes)

                val btnPrev = AppCompatImageButton(searchView.context)
                btnPrev.setImageResource(R.drawable.ic_toolbar_search_prev)
                btnPrev.setBackgroundResource(backgroundRes)

                (searchView.getChildAt(0) as LinearLayout).addView(btnPrev, navButtonsParams)
                (searchView.getChildAt(0) as LinearLayout).addView(btnNext, navButtonsParams)

                btnNext.setOnClickListener { findNext(true) }
                btnPrev.setOnClickListener { findNext(false) }
                searchViewTag++
            }
        }

        val searchManager = activity?.getSystemService(Context.SEARCH_SERVICE) as SearchManager
        searchView.setSearchableInfo(searchManager.getSearchableInfo(activity?.componentName))

        searchView.setIconifiedByDefault(true)
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                findText(newText)
                return false
            }
        })
    }

    private fun findNext(next: Boolean) {
        webView.findNext(next)
    }

    private fun findText(text: String) {
        webView.findAllAsync(text)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        webView.endWork()
    }

    companion object {
        const val ARG_ANNOUNCE_ID = "ARG_ANNOUNCE_ID"
        const val ARG_FORUM_ID = "ARG_FORUM_ID"
        const val JS_INTERFACE = "IAnnounce"
    }
}
